docs(service-worker): improve docs related to `SwUpdate` APIs

gkalpak opened the pull request:
This commit improves documentation related to recently improved or deprecated `SwUpdate` APIs in the following ways:

- Update [check-for-update.service.ts][1] to make use of the return value of [SwUpdate#checkForUpdate()][2].
- Update [prompt-update.service.ts][3] to not call [SwUpdate#activateUpdate()][4] and just reload the page.
- Update the [SwUpdate#activateUpdate()][4] API docs to explain that it is only useful if you want to update a client without reloading and that it can easily lead to version skew.
- Update [a code-snippet][5] to no longer be [hard-coded][6].

[1]: https://github.com/angular/angular/blob/9d9d05911dbd6e2f30e4c7bced0e41fd20ec4285/aio/content/examples/service-worker-getting-started/src/app/check-for-update.service.ts#L16
[2]: https://angular.io/api/service-worker/SwUpdate#checkForUpdate
[3]: https://github.com/angular/angular/blob/9d9d05911dbd6e2f30e4c7bced0e41fd20ec4285/aio/content/examples/service-worker-getting-started/src/app/prompt-update.service.ts#L15
[4]: https://angular.io/api/service-worker/SwUpdate#activateUpdate
[5]: https://github.com/angular/angular/blob/96c6139c9ab35aa6ab2330a5a79a5906d5c2e8be/packages/service-worker/src/update.ts#L44-L54
[6]: https://angular.io/guide/docs-style-guide#hard-coded-snippets

Fixes #43665.

mary-poppins commented:
You can preview 4a7a6cf at https://pr46960-4a7a6cf.ngbuilds.io/.

mary-poppins commented:
You can preview 4742521 at https://pr46960-4742521.ngbuilds.io/.

bob-watson commented on packages/service-worker/src/update.ts:
````diff
@@ -40,18 +40,11 @@ export class SwUpdate {
-   *     available: evt.latestVersion,
-   *   })));
-   * ```
+   * The behavior of `available` can be replicated using `versionUpdates` by filtering for the
````
`* The behavior of `available` can be replicated by using `versionUpdates` by filtering for the`

bob-watson commented on packages/service-worker/src/update.ts:
```diff
@@ -124,6 +117,20 @@ export class SwUpdate {
+   * Only use this method, if you are certain it is safe for your specific use case.
+   *
+   * </div>
+   *
```
This alert seems to be a clearer version of the previous text. Can the preceding text be removed?

bob-watson commented on aio/content/guide/service-worker-communications.md:
```diff
@@ -64,16 +64,21 @@ Alternatively, you might want to define a different [registration strategy](api/
 
-If the current tab needs to be updated to the latest application version immediately, it can ask to do so with the `activateUpdate()` method:
+You can update an existing tab to the latest version by reloading the page as soon as a new version is ready.
+To avoid disrupting the user's progress, it is generally a good idea to prompt the user and let them confirm it is OK to reload the page and update to the latest version:
```
`let them confirm it is OK to reload` ==> `let them confirm that it is OK to reload`

bob-watson commented on aio/content/guide/service-worker-communications.md:
```diff
@@ -64,16 +64,21 @@ Alternatively, you might want to define a different [registration strategy](api/
-Therefore, it is recommended to reload the page once the promise returned by `activateUpdate()` is resolved.
+It is possible to update a tab to the latest version without reloading the page, by calling {@link SwUpdate#activateUpdate SwUpdate#activateUpdate()}.
+
+However, this can easily lead to version skew.
```
A more active voice version is clearer:

`Calling {@link SwUpdate#activateUpdate SwUpdate#activateUpdate()} updates a tab to the latest version without reloading the page, but it could lead to version skew.`

Is `version skew` where a tab is out of sync with the surrounding content in the page? If so, could you say something like that so that understanding the consequences doesn't rely on knowing that term?

bob-watson replied:
This looks good.

re: revision ideas...

Maybe:

```
Calling activateUpdate() updates a tab to the latest version without reloading the page, but this could break the application.
Updating only the tab can create a version mismatch between the [application shell] and other page resources  whose filenames may change between versions. 
```

Could the *lazy-loaded chunk* example be omitted from the alert box since it's mentioned in the following section?

gkalpak replied:
Thx, I split up the sentence as you suggested. I'd rather keep the _lazy-loaded chunks_ example despite it being mentioned in the following section, because the following section in not directly related to this one.

bob-watson reviewed (changes requested):
Looks good.
Just a few minor comments.

mary-poppins commented:
You can preview 0a1087d at https://pr46960-0a1087d.ngbuilds.io/.

NsdHSO commented on aio/content/examples/service-worker-getting-started/src/app/prompt-update.service.ts:
```diff
@@ -1,20 +1,40 @@
 import { Injectable } from '@angular/core';
-import { SwUpdate, UpdateAvailableEvent } from '@angular/service-worker';
+// #docregion sw-replicate-available
+    import { filter, map } from 'rxjs/operators';
```
```suggestion
import { filter, map } from 'rxjs/operators';
```

gkalpak replied:
Thx for pointing this out, @NsdHSO. This is intentional so that the [code-snippet][1] in the [SwUpdate#available API docs][2] is indented uniformly.

This file is not shown in its entirety in the docs (nor is it part of a [live example][3]), therefore it is fine to have inconsistent indentation for the sake of better looking code snippets.

[1]: https://angular.io/guide/docs-style-guide#preparing-source-code-for-code-snippets
[2]: https://pr46960-0a1087d.ngbuilds.io/api/service-worker/SwUpdate#available
[3]: https://angular.io/guide/docs-style-guide#live-examples

mary-poppins commented:
You can preview 04d9ee5 at https://pr46960-04d9ee5.ngbuilds.io/.

josephperrott commented on packages/service-worker/src/update.ts:
```diff
@@ -124,6 +117,20 @@ export class SwUpdate {
    * Updates the current client (i.e. browser tab) to the latest version that is ready for
    * activation.
    *
+   * In most cases, you do not need to use this method and can update a client by reloading the
```
```suggestion
   * In most cases, you should not use this method and instead should update a client by reloading the
```

josephperrott reviewed (approved):
LGTM

gkalpak committed:
docs(service-worker): improve docs related to `SwUpdate` APIs

This commit improves documentation related to recently improved or
deprecated `SwUpdate` APIs in the following ways:

- Update [check-for-update.service.ts][1] to make use of the return
  value of [SwUpdate#checkForUpdate()][2].
- Update [prompt-update.service.ts][3] to not call
  [SwUpdate#activateUpdate()][4] and just reload the page.
- Update the [SwUpdate#activateUpdate()][4] API docs to explain that it
  is only useful if you want to update a client without reloading and
  that it can easily lead to version skew.
- Update [a code-snippet][5] to no longer be [hard-coded][6].

[1]: https://github.com/angular/angular/blob/9d9d05911dbd6e2f30e4c7bced0e41fd20ec4285/aio/content/examples/service-worker-getting-started/src/app/check-for-update.service.ts#L16
[2]: https://angular.io/api/service-worker/SwUpdate#checkForUpdate
[3]: https://github.com/angular/angular/blob/9d9d05911dbd6e2f30e4c7bced0e41fd20ec4285/aio/content/examples/service-worker-getting-started/src/app/prompt-update.service.ts#L15
[4]: https://angular.io/api/service-worker/SwUpdate#activateUpdate
[5]: https://github.com/angular/angular/blob/96c6139c9ab35aa6ab2330a5a79a5906d5c2e8be/packages/service-worker/src/update.ts#L44-L54
[6]: https://angular.io/guide/docs-style-guide#hard-coded-snippets

Fixes #43665

gkalpak committed:
fixup! docs(service-worker): improve docs related to `SwUpdate` APIs

gkalpak committed:
fixup! docs(service-worker): improve docs related to `SwUpdate` APIs

gkalpak committed:
fixup! docs(service-worker): improve docs related to `SwUpdate` APIs

mary-poppins commented:
You can preview f016510 at https://pr46960-f016510.ngbuilds.io/.

AndrewKushnir commented:
This PR was merged into the repository by commit c905da12e2205c8a4bad6b77c0c813fe3eeeb073.
